crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -967,6 +988,9 @@ impl CumulocityConverter {
     ) -> Result<Vec<MqttMessage>, ConversionError> {
         debug!("Mapping message on topic: {}", message.topic.name);
         trace!("Message content: {:?}", message.payload_str());
+
+        record_health_status(&self.mqtt_schema, &mut self.health_status, message);
```

Comment:
Because `process_health_status_message` can address only `Channel::Health` topics, meaning, the topic must have `te/{valid_entity_topic_id}/status/health` scheme. But the requirement is to accept any topic as health endpoint as long as it has `te`(can be custom root) prefix.

Reply:
To support any random topic, that doesn't even have to be a thin-edge topic (not even with custom prefix and custom topic scheme), then we need to make the mapper subscribe to those random topics as well, so that it is aware of the health updates published to those topics. This would mean that we need to dynamically subscribe to those "open topics" as and when they are registered, which won't be easy as our MQTT actor currently doesn't support dynamic subscriptions. The other alternative would be for the mapper to subscribe to '#' and then process only the ones that are declared by the `@health` endpoints. Since the mapper has a persistent connection with the broker, it will unnecessarily increase the load on the broker to queue a lot of unnecessary data for the mapper, most of which won't even be processed.

So, my advice would be to restrict this feature to thin-edge health endpoints only for now (customers are still free to use custom prefixes and topic schemes), and expand it later to any random topics, once we add support for dynamic subscriptions. @reubenmiller Are you okay with that?

Reply:
No the `@health` value will still use the root_prefix (default being `te/`), so at least we need to subscribe for anything more than we are already doing.

The though process was to include the explict `/status/health` portion of the topic is more explicit, rather than having unexpected "status/health" added to the `@health` value.

Reply:
> No the @health value will still use the root_prefix (default being te/), so at least we need to subscribe for anything more than we are already doing.

So, it is guaranteed that the `@health` value would be the health endpoint of some entity, right? It could be using a custom topic scheme, and even a custom prefix, as long as tedge is also configured to use the same prefix. In that case, things are better. The PR description stated that the value `@health` value `can/be/any/custom/topic`, that doesn't even conform to the above mentioned requirements. My concerns were about trying to be so open to any random topic. As long as it is a valid thin-edge health status topic, we're good.

> The though process was to include the explict /status/health portion of the topic is more explicit, rather than having unexpected "status/health" added to the @health value.

I think there's some misunderstanding here. @rina23q implemented this contract with the assumption that we can't assume the `@health` value to be on the `health/status` channel, but rather can be any other random topic as well. If we're restricting the value to the the `health/status` channel, then all good. That's exactly what I was expecting as well.

> Though I guess if someone chooses to uses something like hello/world/status (which will be interpreted as te/hello/world/status, then thin-edge.io won't receive any messages from this topic (as it does not match any of the existing local mqtt topic subscriptions?

Correct. Since the mapper currently subscribes specifically to `te/+/+/+/+/<channel>/<data-type>` topics, Something published to a 4-part topic like `te/hello/world/status` won't even be received, as thin-edge expects at least a 5-part topic.


**TO SUMMARISE** my current understanding is as follows: The requirement was not to support any random topic with any prefix and any number of topic levels (e.g: `a/b/c/d/c/e/f/g/h` or `a/b/c`), but only to expect valid health status messages on explicit health status channel itself, using any valid custom prefix and topic scheme (the same that is configured with tedge). Correct me if this understanding is wrong.

Reply:
Yes the original requirement was to support any random topic (under the root_prefix). But since this opens up the problem that the mapper is not subscribing to a wildcard topic `{root_prefix}/#`, then you're right we won't get the notifications on the topic specified by `@health`.

Then let's change the requirement so that the `@health` should refer to an target entity (e.g. `device/main/sevice/tedge-agent` etc.)...then the agent implies that the `/status/health` topic suffix.

Reply:
> Then let's change the requirement so that the `@health` should refer to an target entity (e.g. `device/main/service/tedge-agent` etc.)...then the agent implies that the `/status/health` topic suffix. This means the `@health` property would follow be of a similar format to the `@parent` field that is used in custom entity registration messages.

I agree that this is the way to go. And not because the former desire to support random topics was adding too many technical constraints on the design.

Having the `@heath` property be an entity identifier makes sense and even more sense than a random topic:
- even if a child device is not running `tedge-agent`, this child device will have in any cases to define and populate `te` topics for its data and services: hence this is not restrictive
- the  `@heath` property is the identifier of the entity, more likely a service but not necessarily,  used to monitor the health of the device: this makes sense from a user perspective
- this is aligned to all the other proposals related to MQTT topics: easier to understand and implement.